src/ripple/shamap/README.md

```diff
@@ -26,164 +45,322 @@ There are two different ways of building and using a SHAMap:
+    does that work for us.
 
  3. Finally, if ledgerSeq_ is non-zero and we did't locate the node in the
     historic data, then we call a MissingNodeHandler.
 
     The non-zero ledgerSeq_ indicates that the SHAMap is a complete map that
-    belongs to a historic ledger with the given (non-zero) sequence number.
-    So, if all expected data is always present, the MissingNodeHandler should
-    never be executed.
+    belongs to a historic ledger with the given (non-zero, unsharable) sequence
+    number. So, if all expected data is always present, the MissingNodeHandler
+    should never be executed.
 
-    And, since we now know that this SHAMap does not fully represent
-    the data from that ledger, we set the SHAMap's sequence number to zero.
+    And, since we now know that this SHAMap does not fully represent the data
+    from that ledger, we set the SHAMap's sequence number to zero.
 
 If phase 1 returned a node, then we already know that the node is immutable.
 However, if either phase 2 executes successfully, then we need to turn the
 returned node into an immutable node.  That's handled by the call to
-`make_shared<SHAMapTreeNode>` inside the try block.  That code is inside
-a try block because the `fetchNodeExternalNT` method promises not to throw.
-In case the constructor called by `make_shared` throws we don't want to
-break our promise.
+`make_shared<SHAMapTreeNode>` inside the try block.  That code is inside a try
+block because the `fetchNodeExternalNT` method promises not to throw. In case
+the constructor called by `make_shared` throws we don't want to break our
+promise.
 
 
 ## Canonicalize ##
 
-The calls to `canonicalize()` make sure that if the resulting node is already
-in the SHAMap, then we return the node that's already present -- we never
-replace a pre-existing node.  By using `canonicalize()` we manage a thread
-race condition where two different threads might both recognize the lack of a
-SHAMapTreeNode at the same time.  If they both attempt to insert the node
-then `canonicalize` makes sure that the first node in wins and the slower
-thread receives back a pointer to the node inserted by the faster thread.
+`canonicalize()` is called every time a node is introduced into the SHAMap.
+
+A call to `canonicalize()` stores the node in the `TreeNodeCache` if it does not
+already exist in the `TreeNodeCache`.
+
+The calls to `canonicalize()` make sure that if the resulting node is already in
+the SHAMap, node `TreeNodeCache` or database, then we don't create duplicates.
+
+On retrieving existing nodes, the copy in the database has top priority.  If
+there's a duplicate in the `TreeNodeCache`, it is replaced by the copy in the
+database. Or if there is a duplicate found between the `TreeNodeCache` and the
+SHAMap, the copy in the `TreeNodeCache` prevails and replaces that in the
+SHAMap.
+
+On inserting new nodes (such as new transactions or account states), the
```

Comment:
A bit confused here. When we insert a new node, how could there be a copy of that node, since the node we are inserting is new?

Reply:
Rewritten.  This part was incorrect.